contracts/token/ERC20Rebasable.sol

```diff
@@ -0,0 +1,358 @@
+    /// @param to_  An address of the recipient of the tokens
+    /// @param amount_ An amount of tokens to transfer
+    function _transfer(
+        address from_,
+        address to_,
+        uint256 amount_
+    ) internal onlyNonZeroAccount(from_) onlyNonZeroAccount(to_) {
+        uint256 sharesToTransfer = _getSharesByTokens(amount_);
+        _transferShares(from_, to_, sharesToTransfer);
+        _emitTransferEvents(from_, to_, amount_ ,sharesToTransfer);
+    }
+
+    /// @dev Updates owner_'s allowance for spender_ based on spent amount_. Does not update
+    ///     the allowance amount in case of infinite allowance
+    /// @param owner_ An address of the account to spend allowance
+    /// @param spender_ An address of the spender of the tokens
+    /// @param amount_ An amount of allowance spend
+    function _spendAllowance(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal {
+        uint256 currentAllowance = _getTokenAllowance()[owner_][spender_];
+        if (currentAllowance == type(uint256).max) {
+            return;
+        }
+        if (amount_ > currentAllowance) {
+            revert ErrorNotEnoughAllowance();
+        }
+        unchecked {
+            _approve(owner_, spender_, currentAllowance - amount_);
+        }
+    }
+
+    /// @dev Sets amount_ as the allowance of spender_ over the owner_'s tokens
+    /// @param owner_ An address of the account to set allowance
+    /// @param spender_ An address of the tokens spender
+    /// @param amount_ An amount of tokens to allow to spend
+    function _approve(
+        address owner_,
+        address spender_,
+        uint256 amount_
+    ) internal virtual onlyNonZeroAccount(owner_) onlyNonZeroAccount(spender_) {
+        _getTokenAllowance()[owner_][spender_] = amount_;
+        emit Approval(owner_, spender_, amount_);
+    }
+
+    function _sharesOf(address account_) internal view returns (uint256) {
+        return _getShares()[account_];
+    }
+
+    function _getTokensByShares(uint256 sharesAmount_) internal view returns (uint256) {
+        (uint256 tokensRate, uint256 decimals) = _getTokenRateAndDecimal();
+        return (sharesAmount_ * tokensRate) / (10 ** decimals);
+    }
+
+    function _getSharesByTokens(uint256 tokenAmount_) internal view returns (uint256) {
+        (uint256 tokensRate, uint256 decimals) = _getTokenRateAndDecimal();
+        return (tokenAmount_ * (10 ** decimals)) / tokensRate;
+    }
+
+    function _getTokenRateAndDecimal() internal view returns (uint256, uint256) {
+        uint8 rateDecimals = TOKEN_RATE_ORACLE.decimals();
+
+        if (rateDecimals == uint8(0)) revert ErrorTokenRateDecimalsIsZero();
+
+        //slither-disable-next-line unused-return
+        (,
+        int256 answer
+        ,
+        ,
+        uint256 updatedAt
+        ,) = TOKEN_RATE_ORACLE.latestRoundData();
+
+        if (updatedAt == 0) revert ErrorWrongOracleUpdateTime();
```

Comment:
:warning: :red_circle:  here we can once again prevent token from minting on L2 when it was bridged from L1
the solution space is:
- accept the risks
- mint wstETH instead of stETH

let's outline it in the spec and ask reviewers to weigh in